### ISSUE TITLE
the tricky bug on hexCommand

### ISSUE THREAD

**Issue by aliokan** [HaxeFoundation/hashlink#154]
@ncannasse Is the one I told you during the Summit. 

I am not able to reproduce it without the framework. 
You can found the project source, dump files and hl generated on the link: https://drive.google.com/open?id=1Jyue2an6kHTNilGbEG0uMwcJI92tQ6s3 
I use haxe nightly version and a build of hashlink made this morning with the git version, on Mac 10.13.4 (17E202)

to compile and run the project: `haxe build-hl.hxml && hl bin/test.hl`

### logs:
```
hex/control/trigger/mock/MockMacroCommand.hx:49: MockMacroCommand new
hex/di/Injector.hx:157: instance: #hex.control.trigger.mock.MockMacroCommand
hex/di/Injector.hx:158: instance.acceptInjector: function#10DAA1780
hex/di/Injector.hx:363: target: #hex.control.trigger.mock.MockMacroCommand
hex/di/Injector.hx:364: acceptInjector BEFORE
hex/di/annotation/FastInjectionBuilder.hx:42: applyInjection START: hex.control.trigger.Command
hex/di/annotation/FastInjectionBuilder.hx:150: applyInjection END: hex.control.trigger.Command
hex/di/Injector.hx:366: acceptInjector AFTER
```
### What happen: 
- `MainMachinaTest` line 27: `emu.addTestMethod(MacroCommandTest, "testMacroCommandExecution");`
...
- `MacroCommandTest` line 13: `var macroCommand = injector.getOrCreateNewInstance( MockMacroCommand );`
- `hex.di.Injector` line 180: `return this.satisfies( type ) ? this.getInstance( type ) : this.instantiateUnmapped( type );`
- `hex.di.Injector` line 161: `this._applyInjection( instance, type );`
- `hex.di.Injector` line 365: `target.acceptInjector( this );`

`target` is an instance of `hex.control.trigger.mock.MockMacroCommand` descendant of `hex.control.trigger.MacroCommand` and `hex.control.trigger.Command`. 
that class contains an overridden version of acceptInjector, but on hl the call of `acceptInjector` is made on his grandparent `hex.control.trigger.Command` instead of `hex.control.trigger.mock.MockMacroCommand`. 
 
`acceptInjector` methods are generated by Macro in `hex.di.annotation.FastInjectionBuilder` line 151. 

When I check the dump files, `hex.control.trigger.mock.MockMacroCommand` contains `acceptInjector` method.

**Comment by FrancisBourre** [comment]
Hello, by chance, any news about this one?

**Comment by ncannasse** [comment]
Didn't have time to look at it yet, but will soon. Does it involve some Reflect.callMethod ? so far that's the main cause for compatibility issue between HL and other Haxe targets, so I want to add more unit tests for it to uniformise behaviors in Haxe 4.0

**Comment by ncannasse** [comment]
Related https://github.com/HaxeFoundation/haxe/issues/7106

**Comment by aliokan** [comment]
Are you sure is related to reflection? 

Bellow, the generated code: 
`applyInjection START` and `applyInjection END` in the output, but not `MacroCommand preExecute`. Look like `this.preExecute` is removed during HL compilation, at lease the method i never executed. 

### in hex.control.trigger.mock.MockMacroCommand.dump 
```haxe
@:noCompletion @:keep
public function acceptInjector(i:hex.di.IBasicInjector) {
	var t = hex.control.trigger.mock.MockMacroCommand;
	var f = i.getInstanceWithClassName;
	haxe.Log.trace("applyInjection START: " + "hex.control.trigger.mock.MockMacroCommand", {fileName : "hex/di/annotation/FastInjectionBuilder.hx", lineNumber : 42, className : "hex.control.trigger.mock.MockMacroCommand", methodName : "acceptInjector"});
	var this1 = "hex.log.ILogger";
	var this2 = "";
	this.logger = f(cast this1, @:implicitCast cast this2, t, false);
	var this3 = "hex.di.IDependencyInjector";
	var this4 = "";
	var this5 = "Array<hex.control.payload.ExecutionPayload>";
	var this6 = "";
	this.preExecute(f(cast this3, @:implicitCast cast this4, null, true), f(cast this5, @:implicitCast cast this6, null, true));
	haxe.Log.trace("applyInjection END: " + "hex.control.trigger.mock.MockMacroCommand", {fileName : "hex/di/annotation/FastInjectionBuilder.hx", lineNumber : 150, className : "hex.control.trigger.mock.MockMacroCommand", methodName : "acceptInjector"});
}
```

### in hex.control.trigger.MacroCommand.dump
```haxe
public function preExecute(injector:hex.di.IDependencyInjector, payloads:Array<hex.control.payload.ExecutionPayload>) {
	haxe.Log.trace("MacroCommand preExecute", {fileName : "hex/control/trigger/MacroCommand.hx", lineNumber : 41, className : "hex.control.trigger.MacroCommand", methodName : "preExecute"});
	this._injector = injector;
	this._payloads = payloads;
	this._prepare();
}
```

**Comment by aliokan** [comment]
Now, when I'm trying to build the example project with the nightly build `1e3e5e0`, I get these errors during compilation: 
```
.haxelib/hexannotation/git/src/hex/di/annotation/FastInjectionBuilder.hx:33: --- hex.control.trigger.Command
.haxelib/hexannotation/git/src/hex/di/annotation/FastInjectionBuilder.hx:33: --- hex.control.trigger.MacroCommand
.haxelib/hexannotation/git/src/hex/di/annotation/FastInjectionBuilder.hx:33: --- hex.control.trigger.mock.MockMacroCommand
/Users/user/haxe/versions/1e3e5e0/std/haxe/macro/Expr.hx:790: characters 2-10 : Duplicate access modifier public
.haxelib/hexcommand/git/src/hex/control/trigger/Command.hx:16: lines 16-147 : Defined in this class
/Users/user/haxe/versions/1e3e5e0/std/haxe/macro/Expr.hx:790: characters 2-10 : Previously defined here
.haxelib/hexcommand/git/src/hex/control/trigger/Command.hx:16: lines 16-147 : Defined in this class
/Users/user/haxe/versions/1e3e5e0/std/haxe/macro/Expr.hx:790: characters 2-10 : Duplicate access modifier public
.haxelib/hexcommand/git/src/hex/control/trigger/MacroCommand.hx:16: lines 16-212 : Defined in this class
/Users/user/haxe/versions/1e3e5e0/std/haxe/macro/Expr.hx:790: characters 2-10 : Previously defined here
.haxelib/hexcommand/git/src/hex/control/trigger/MacroCommand.hx:16: lines 16-212 : Defined in this class
/Users/user/haxe/versions/1e3e5e0/std/haxe/macro/Expr.hx:790: characters 2-10 : Duplicate access modifier public
.haxelib/hexcommand/git/test/hex/control/trigger/mock/MockMacroCommand.hx:10: lines 10-62 : Defined in this class
/Users/user/haxe/versions/1e3e5e0/std/haxe/macro/Expr.hx:790: characters 2-10 : Previously defined here
.haxelib/hexcommand/git/test/hex/control/trigger/mock/MockMacroCommand.hx:10: lines 10-62 : Defined in this class
```

**Comment by ncannasse** [comment]
I'm fine changing with that new behavior for Haxe 4.0, it should be pretty
rare

**Comment by aliokan** [comment]
I still have the `Duplicate access modifier public` error with the current nightly build `bdccacc`.

**Comment by aliokan** [comment]
How is it possible to generate `public public`?

**Comment by Gama11** [comment]
Well, `access` is just an `Array`, so that's quite easy to do:

```haxe
import haxe.macro.Context;
import haxe.macro.Expr;

class Macro {
	macro static public function build():Array<Field> {
		var fields = Context.getBuildFields();
		var newField = {
			name: "foo",
			doc: null,
			meta: [],
			access: [AStatic, APublic, APublic],
			kind: FVar(macro:String, macro "bar"),
			pos: Context.currentPos()
		};
		fields.push(newField);
		return fields;
	}
}
```

```haxe
@:build(Macro.build())
class Main {
	public static function main() {}
}
````

```
haxe\std/haxe/macro/Expr.hx:790: characters 2-10 : Duplicate access modifier public
source/Main.hx:2: lines 2-6 : Defined in this class
haxe\std/haxe/macro/Expr.hx:790: characters 2-10 : Previously defined here
```

**Comment by ncannasse** [comment]
@aliokan any update regarding this issue ? Could you reproduce with latest Haxe build ?

**Comment by aliokan** [comment]
Actually, we just get `Not_found` message on all targets when we compile hexDSL with the current nightly build. https://travis-ci.org/DoclerLabs/hexDSL/builds/436104227

**Comment by ncannasse** [comment]
Uh-oh, @simn can you look into this?

**Comment by Simn** [comment]
Any instructions how to reproduce this?

**Comment by aliokan** [comment]
```
git clone https://github.com/DoclerLabs/hexDSL.git hexdsl
cd hexdsl/
haxelib newrepo
haxelib install haxelib.json --always
haxelib git hexunit https://github.com/DoclerLabs/hexUnit.git
haxelib dev hexdsl .
haxe build-neko.hxml && neko bin/MainTest.n
```

**Comment by Simn** [comment]
Something looks wrong here: https://i.imgur.com/YAwKQba.png

The compiler thinks it's a MapVO, but it's actually a ConstructorVO at run-time and then it can't find the `getPropertyKey` field on that.

I'll certainly have to improve the error message though.

**Comment by Simn** [comment]
I suppose you're relying on this raising a catchable exception. I've made it raise such an exception on latest dev again, so the compilation works.

You might still want to look at that code though.

**Comment by FrancisBourre** [comment]
Thank you so much @Simn !
Yeah, it's because of [this temporary hack](https://github.com/DoclerLabs/hexDSL/blob/master/src/hex/compiletime/factory/ExpressionFactory.hx#L25) I forgot to remove.
Can fix it now, and see if we got the previous error. Thanks again.

**Comment by ncannasse** [comment]
@FrancisBourre up ? :) with Haxe GIT or Preview 5

**Comment by ncannasse** [comment]
I think I might have found something related to this error. In some very specific cases, a dynamic call might have been calling a method on a base class instead of its overriden variant. This has just been fixed and will be part of HL 1.9 release, please tell me if you can test with GIT

**Comment by ncannasse** [comment]
Last call before issue close :)
